File: apps/web/src/lib/bots.ts
import { api, APIError } from "./api";
import type { User, Workspace } from "./types";

export type BotToken = {
  id: string;
  bot_user_id: string;
  workspace_id: string;
  owner_user_id?: string;
  name: string;
  scopes: string[];
  created_by?: string;
  created_at: string;
  last_used_at?: string;
  revoked_at?: string;
  token?: string;
};

export type BotWithTokens = {
  bot: User;
  tokens: BotToken[];
};

export type DeletedBot = {
  id: string;
  display_name: string;
  former_handle: string;
  deleted_at: string;
};

export type OwnedBotWorkspace = {
  id: string;
  route_id: string;
  name: string;
};

export type OwnedBotEntry = {
  bot: User;
  workspace: OwnedBotWorkspace;
  active_token_count: number;
};

export type BotScopeBundle = "bot:read" | "bot:write" | "bot:admin";

export const BOT_SCOPE_BUNDLES: { id: BotScopeBundle; label: string; hint: string }[] = [
  {
    id: "bot:read",
    label: "Read",
    hint: "View channels, messages, and threads. No write access.",
  },
  {
    id: "bot:write",
    label: "Read & write",
    hint: "Post and edit messages, send DMs, upload attachments, and publish command menus.",
  },
  {
    id: "bot:admin",
    label: "Admin",
    hint: "Read & write, publish command menus, and manage channels. Use sparingly.",
  },
];

export type CreateBotInput = {
  display_name: string;
  handle?: string;
  avatar_url?: string;
  owner_user_id?: string;
  token_name?: string;
  scopes?: string[];
  setup_nonce?: string;
  // false skips the initial token mint (setup-code installs mint the
  // token at claim time instead).
  initial_token?: boolean;
};

export type CreateBotResponse = {
  bot: User;
  // Omitted when the bot was created with initial_token: false.
  bot_token?: BotToken;
};

export type BotSetupCode = {
  id: string;
  bot_user_id: string;
  workspace_id: string;
  token_name: string;
  scopes: string[];
  defaults: BotSetupCodeDefaults;
  created_by?: string;
  created_at: string;
  expires_at: string;
  // One-time plaintext code (XXXX-XXXX-XXXX). Present only in the mint
  // response; the server stores just a hash.
  code?: string;
};

export type BotSetupCodeDefaults = {
  defaultTo?: string;
  allowFrom?: string[];
  agentActivity?: boolean;
};

export async function listWorkspaceBots(workspaceID: string): Promise<BotWithTokens[]> {
  const data = await api<{ bots: BotWithTokens[] }>(`/api/workspaces/${workspaceID}/bots`);
  return data.bots ?? [];
}

export async function createWorkspaceBot(
  workspaceID: string,
  input: CreateBotInput,
): Promise<CreateBotResponse> {
  return api<CreateBotResponse>(`/api/workspaces/${workspaceID}/bots`, {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function listWorkspaceBotTokens(
  workspaceID: string,
  botUserID: string,
): Promise<BotToken[]> {
  const data = await api<{ bot_tokens: BotToken[] }>(
    `/api/workspaces/${workspaceID}/bots/${botUserID}/tokens`,
  );
  return data.bot_tokens ?? [];
}

export async function createWorkspaceBotToken(
  workspaceID: string,
  botUserID: string,
  input: { name?: string; scopes?: string[]; setup_nonce?: string },
): Promise<BotToken> {
  const data = await api<{ bot_token: BotToken }>(
    `/api/workspaces/${workspaceID}/bots/${botUserID}/tokens`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
  return data.bot_token;
}

export async function createWorkspaceBotSetupCode(
  workspaceID: string,
  botUserID: string,
  input: { name?: string; scopes?: string[]; defaults?: BotSetupCodeDefaults },
): Promise<BotSetupCode> {
  const data = await api<{ setup_code: BotSetupCode }>(
    `/api/workspaces/${workspaceID}/bots/${botUserID}/setup-codes`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
  return data.setup_code;
}

export async function listBotTokens(botUserID: string): Promise<BotToken[]> {
  const data = await api<{ bot_tokens: BotToken[] }>(`/api/bots/${botUserID}/tokens`);
  return data.bot_tokens ?? [];
}

export async function createBotToken(
  botUserID: string,
  input: { name?: string; scopes?: string[]; setup_nonce?: string },
): Promise<BotToken> {
  const data = await api<{ bot_token: BotToken }>(`/api/bots/${botUserID}/tokens`, {
    method: "POST",
    body: JSON.stringify(input),
  });
  return data.bot_token;
}

export async function revokeBotToken(tokenID: string): Promise<BotToken> {
  const data = await api<{ bot_token: BotToken }>(`/api/bot-tokens/${tokenID}/revoke`, {
    method: "POST",
    body: JSON.stringify({}),
  });
  return data.bot_token;
}

export async function removeBotFromWorkspace(
  workspaceID: string,
  botUserID: string,
): Promise<void> {
  await api(`/api/workspaces/${workspaceID}/bots/${botUserID}/membership`, {
    method: "DELETE",
  });
}

export async function deleteBot(botUserID: string): Promise<DeletedBot> {
  const data = await api<{ deleted_bot: DeletedBot }>(`/api/bots/${botUserID}`, {
    method: "DELETE",
  });
  return data.deleted_bot;
}

export async function listMyBots(): Promise<OwnedBotEntry[]> {
  const data = await api<{ bots: OwnedBotEntry[] }>("/api/me/bots");
  return data.bots ?? [];
}

export function botLoadErrorMessage(err: unknown): string {
  if (err instanceof APIError) {
    if (err.status === 401) return "Sign in to manage bots.";
    if (err.status === 403) return "You don't have permission to manage bots in this workspace.";
    if (err.status === 404) return "That bot or workspace is no longer available.";
    if (err.status === 409) return "That handle is already taken. Try another.";
    if (err.status === 400) return err.message || "That request is invalid.";
  }
  return err instanceof Error ? err.message : "Something went wrong";
}

export function isServiceBot(bot: { owner_user_id?: string }): boolean {
  return !bot.owner_user_id;
}

export function activeTokens(tokens: BotToken[] | undefined): BotToken[] {
  if (!tokens) return [];
  return tokens.filter((t) => !t.revoked_at);
}

export function suggestHandleFrom(displayName: string): string {
  return displayName
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 32);
}

export type OpenClawAccountMode = "single" | "named";

export function openClawWorkspaceIdentifier(workspace: Pick<Workspace, "id" | "slug">): string {
  return workspace.slug.trim() || workspace.id;
}

function jsonString(value: string): string {
  return JSON.stringify(value);
}

function envNameForHandle(handle: string): string {
  const suffix = handle
    .replace(/^@/, "")
    .toUpperCase()
    .replace(/[^A-Z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "");
  return suffix ? `CLICKCLACK_${suffix}_BOT_TOKEN` : "CLICKCLACK_BOT_TOKEN";
}

function shellQuote(value: string): string {
  return `'${value.replaceAll("'", `'"'"'`)}'`;
}

export function buildOpenClawConfigSnippet(opts: {
  workspace: string;
  botHandle: string;
  botUserID: string;
  mode: OpenClawAccountMode;
  baseURL?: string;
  defaultTo?: string;
  allowFrom?: string[];
  agentActivity?: boolean;
}): string {
  const base = (
    opts.baseURL || (typeof window !== "undefined" ? window.location.origin : "")
  ).replace(/\/$/, "");
  const handle = opts.botHandle.replace(/^@/, "");
  const envName = opts.mode === "single" ? "CLICKCLACK_BOT_TOKEN" : envNameForHandle(handle);
  const baseURL = base || "https://your-clickclack.example.com";
  const defaultTo = opts.defaultTo?.trim() || "channel:general";

  // Account-level lines shared by both single and named-account shapes.
  const accountLines = (indent: string): string => {
    const lines = [
      `workspace: ${jsonString(opts.workspace)},`,
      `botUserId: ${jsonString(opts.botUserID)},`,
      `defaultTo: ${jsonString(defaultTo)},`,
    ];
    if (opts.allowFrom && opts.allowFrom.length > 0 && !opts.allowFrom.includes("*")) {
      lines.push(`allowFrom: [${opts.allowFrom.map(jsonString).join(", ")}],`);
    }
    if (opts.agentActivity) {
      lines.push(`agentActivity: true,`);
    }
    return lines.map((line) => indent + line).join("\n");
  };

  if (opts.mode === "named") {
    return `{
  channels: {
    clickclack: {
      enabled: true,
      baseUrl: ${jsonString(baseURL)},
      defaultAccount: ${jsonString(handle)},
      accounts: {
        ${jsonString(handle)}: {
          token: { source: "env", provider: "default", id: ${jsonString(envName)} },
${accountLines("          ")}
        },
      },
    },
  },
}`;
  }

  return `{
  channels: {
    clickclack: {
      enabled: true,
      baseUrl: ${jsonString(baseURL)},
      token: { source: "env", provider: "default", id: ${jsonString(envName)} },
${accountLines("      ")}
    },
  },
}`;
}

export function buildOpenClawCodeSnippet(opts: {
  code: string;
  botHandle: string;
  mode: OpenClawAccountMode;
  baseURL?: string;
}): string {
  const base =
    (opts.baseURL || (typeof window !== "undefined" ? window.location.origin : "")).replace(
      /\/$/,
      "",
    ) || "https://your-clickclack.example.com";
  const handle = opts.botHandle.replace(/^@/, "");
  const accountArg = opts.mode === "named" ? ` --account ${shellQuote(handle)}` : "";
  return `openclaw channels add clickclack${accountArg} --code ${shellQuote(`${base}/#${opts.code}`)}`;
}

export function buildOpenClawShellSnippet(opts: {
  botHandle: string;
  token: string;
  mode: OpenClawAccountMode;
  workspace: string;
  baseURL?: string;
}): string {
  const base =
    (opts.baseURL || (typeof window !== "undefined" ? window.location.origin : "")).replace(
      /\/$/,
      "",
    ) || "https://your-clickclack.example.com";
  const handle = opts.botHandle.replace(/^@/, "");
  const accountLine = opts.mode === "named" ? ` \\\n  --account ${shellQuote(handle)}` : "";
  return `openclaw channels add clickclack${accountLine} \\
  --base-url ${shellQuote(base)} \\
  --token ${shellQuote(opts.token)} \\
  --workspace ${shellQuote(opts.workspace)}`;
}

File: apps/api/internal/httpapi/bots.go
package httpapi

import (
	"database/sql"
	"errors"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/openclaw/clickclack/apps/api/internal/store"
)

func (s *Server) listBots(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot manage bots"))
		return
	}
	bots, err := s.store.ListBots(r.Context(), chi.URLParam(r, "workspace_id"), act.user.ID)
	writeResult(w, map[string]any{"bots": bots}, err)
}

func (s *Server) listMyBots(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot list owned bots"))
		return
	}
	bots, err := s.store.ListBotsOwnedBy(r.Context(), act.user.ID)
	writeResult(w, map[string]any{"bots": bots}, err)
}

func (s *Server) listBotCommands(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	workspaceID := chi.URLParam(r, "workspace_id")
	if err := act.requireScope("workspaces:read"); err != nil {
		writeError(w, http.StatusForbidden, err)
		return
	}
	if err := act.requireWorkspace(workspaceID); err != nil {
		writeError(w, http.StatusForbidden, err)
		return
	}
	commands, err := s.store.ListBotCommands(r.Context(), workspaceID, act.user.ID)
	if errors.Is(err, sql.ErrNoRows) {
		writeError(w, http.StatusForbidden, errors.New("workspace membership required"))
		return
	}
	writeResult(w, map[string]any{"bot_commands": commands}, err)
}

func (s *Server) setBotCommands(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID == "" {
		writeError(w, http.StatusForbidden, errors.New("bot token required"))
		return
	}
	if err := act.requireScope(store.BotCommandsWriteScope); err != nil {
		writeError(w, http.StatusForbidden, err)
		return
	}
	var body struct {
		Commands *[]store.BotCommandInput `json:"commands"`
	}
	if err := readJSON(w, r, &body); err != nil {
		writeError(w, http.StatusBadRequest, err)
		return
	}
	if body.Commands == nil {
		writeError(w, http.StatusBadRequest, errors.New("commands is required"))
		return
	}
	commands, err := s.store.SetBotCommands(r.Context(), act.workspaceID, act.user.ID, *body.Commands)
	if err != nil {
		writeStoreError(w, err)
		return
	}
	s.publishEvent(r.Context(), store.Event{
		Type:        "bot_command.updated",
		WorkspaceID: act.workspaceID,
		CreatedAt:   time.Now().UTC().Format(time.RFC3339Nano),
		Payload: map[string]string{
			"workspace_id": act.workspaceID,
			"bot_user_id":  act.user.ID,
		},
	})
	writeJSON(w, http.StatusOK, map[string]any{"bot_commands": commands})
}

func (s *Server) createBot(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot create bots"))
		return
	}
	var body struct {
		OwnerUserID  string   `json:"owner_user_id"`
		DisplayName  string   `json:"display_name"`
		Handle       string   `json:"handle"`
		AvatarURL    string   `json:"avatar_url"`
		TokenName    string   `json:"token_name"`
		Scopes       []string `json:"scopes"`
		SetupNonce   string   `json:"setup_nonce"`
		InitialToken *bool    `json:"initial_token"`
	}
	if err := readJSON(w, r, &body); err != nil {
		writeError(w, http.StatusBadRequest, err)
		return
	}
	skipInitialToken := body.InitialToken != nil && !*body.InitialToken
	bot, token, err := s.store.CreateBot(r.Context(), store.CreateBotInput{
		WorkspaceID:      chi.URLParam(r, "workspace_id"),
		OwnerUserID:      body.OwnerUserID,
		DisplayName:      body.DisplayName,
		Handle:           body.Handle,
		AvatarURL:        body.AvatarURL,
		TokenName:        body.TokenName,
		Scopes:           body.Scopes,
		SetupNonce:       body.SetupNonce,
		CreatedBy:        act.user.ID,
		SkipInitialToken: skipInitialToken,
	})
	result := map[string]any{"bot": bot}
	if !skipInitialToken {
		result["bot_token"] = token
	}
	writeResultStatus(w, http.StatusCreated, result, err)
}

func (s *Server) listBotTokens(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot manage bot tokens"))
		return
	}
	tokens, err := s.store.ListBotTokens(r.Context(), chi.URLParam(r, "bot_user_id"), act.user.ID)
	writeResult(w, map[string]any{"bot_tokens": tokens}, err)
}

func (s *Server) listWorkspaceBotTokens(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot manage bot tokens"))
		return
	}
	tokens, err := s.store.ListBotTokensForWorkspace(r.Context(), chi.URLParam(r, "workspace_id"), chi.URLParam(r, "bot_user_id"), act.user.ID)
	writeResult(w, map[string]any{"bot_tokens": tokens}, err)
}

func (s *Server) createBotToken(w http.ResponseWriter, r *http.Request) {
	s.createBotTokenForWorkspace(w, r, "")
}

func (s *Server) createWorkspaceBotToken(w http.ResponseWriter, r *http.Request) {
	s.createBotTokenForWorkspace(w, r, chi.URLParam(r, "workspace_id"))
}

func (s *Server) createBotTokenForWorkspace(w http.ResponseWriter, r *http.Request, workspaceID string) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot create bot tokens"))
		return
	}
	var body struct {
		Name       string   `json:"name"`
		Scopes     []string `json:"scopes"`
		SetupNonce string   `json:"setup_nonce"`
	}
	if err := readJSON(w, r, &body); err != nil {
		writeError(w, http.StatusBadRequest, err)
		return
	}
	token, err := s.store.CreateBotToken(r.Context(), store.CreateBotTokenInput{
		WorkspaceID: workspaceID,
		BotUserID:   chi.URLParam(r, "bot_user_id"),
		Name:        body.Name,
		Scopes:      body.Scopes,
		SetupNonce:  body.SetupNonce,
		CreatedBy:   act.user.ID,
	})
	writeResultStatus(w, http.StatusCreated, map[string]any{"bot_token": token}, err)
}

func (s *Server) createWorkspaceBotSetupCode(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot create bot setup codes"))
		return
	}
	var body struct {
		Name     string                     `json:"name"`
		Scopes   []string                   `json:"scopes"`
		Defaults store.BotSetupCodeDefaults `json:"defaults"`
	}
	if err := readJSON(w, r, &body); err != nil {
		writeError(w, http.StatusBadRequest, err)
		return
	}
	code, err := s.store.CreateBotSetupCode(r.Context(), store.CreateBotSetupCodeInput{
		WorkspaceID: chi.URLParam(r, "workspace_id"),
		BotUserID:   chi.URLParam(r, "bot_user_id"),
		Name:        body.Name,
		Scopes:      body.Scopes,
		Defaults:    body.Defaults,
		CreatedBy:   act.user.ID,
	})
	if err == nil {
		s.recordAudit(r.Context(), code.WorkspaceID, act.user.ID, "bot_setup_code.created", "bot_setup_code", code.ID, map[string]any{
			"bot_user_id": code.BotUserID,
			"token_name":  code.TokenName,
		})
	}
	if err != nil {
		writeStoreError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, map[string]any{"setup_code": s.setupCodeContract(r, code)})
}

func (s *Server) claimBotSetupCode(w http.ResponseWriter, r *http.Request) {
	if !s.setupCodeClaimLimiter.allow(clientIPKey(r)) {
		writeError(w, http.StatusTooManyRequests, errors.New("too many setup code attempts, retry later"))
		return
	}
	var body struct {
		Code string `json:"code"`
	}
	if err := readJSON(w, r, &body); err != nil {
		writeError(w, http.StatusBadRequest, err)
		return
	}
	claim, err := s.store.ClaimBotSetupCode(r.Context(), body.Code)
	if err != nil {
		writeStoreError(w, err)
		return
	}
	s.recordAudit(r.Context(), claim.Workspace.ID, claim.Bot.ID, "bot_setup_code.claimed", "bot_token", claim.BotToken.ID, map[string]any{
		"bot_user_id": claim.Bot.ID,
		"token_name":  claim.BotToken.Name,
	})
	response := map[string]any{
		"token": claim.BotToken.Token,
		"bot": map[string]string{
			"id":           claim.Bot.ID,
			"handle":       claim.Bot.Handle,
			"display_name": claim.Bot.DisplayName,
		},
		"workspace": map[string]string{
			"id":       claim.Workspace.ID,
			"route_id": claim.Workspace.RouteID,
			"slug":     claim.Workspace.Slug,
			"name":     claim.Workspace.Name,
		},
		"defaults": claim.Defaults,
	}
	if baseURL := s.apiBaseURL(r); baseURL != "" {
		response["contract_version"] = botSetupContractVersion
		response["api_base_url"] = baseURL
	}
	writeJSON(w, http.StatusOK, response)
}

const botSetupContractVersion = 1

type botSetupCodeContract struct {
	store.BotSetupCode
	ContractVersion int    `json:"contract_version,omitempty"`
	ClaimURL        string `json:"claim_url,omitempty"`
	APIBaseURL      string `json:"api_base_url,omitempty"`
}

func (s *Server) setupCodeContract(r *http.Request, code store.BotSetupCode) botSetupCodeContract {
	response := botSetupCodeContract{BotSetupCode: code}
	if baseURL := s.apiBaseURL(r); baseURL != "" {
		response.ContractVersion = botSetupContractVersion
		response.APIBaseURL = baseURL
		response.ClaimURL = baseURL + "/api/bot-setup-codes/claim"
	}
	return response
}

func (s *Server) removeBotFromWorkspace(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot remove bots from workspaces"))
		return
	}
	err = s.store.RemoveBotFromWorkspace(r.Context(), chi.URLParam(r, "workspace_id"), chi.URLParam(r, "bot_user_id"), act.user.ID)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			writeError(w, http.StatusNotFound, err)
			return
		}
		writeStoreError(w, err)
		return
	}
	s.publishBotMembershipRemoved(chi.URLParam(r, "workspace_id"), chi.URLParam(r, "bot_user_id"))
	w.WriteHeader(http.StatusNoContent)
}

func (s *Server) deleteBot(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot delete bots"))
		return
	}
	deleted, err := s.store.DeleteBot(r.Context(), chi.URLParam(r, "bot_user_id"), act.user.ID)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			writeError(w, http.StatusNotFound, err)
			return
		}
		writeStoreError(w, err)
		return
	}
	s.publishBotDeleted(deleted)
	writeJSON(w, http.StatusOK, map[string]any{"deleted_bot": deleted})
}

func (s *Server) revokeBotToken(w http.ResponseWriter, r *http.Request) {
	act, err := s.currentActor(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, err)
		return
	}
	if act.botTokenID != "" {
		writeError(w, http.StatusForbidden, errors.New("bot tokens cannot revoke bot tokens"))
		return
	}
	token, err := s.store.RevokeBotToken(r.Context(), chi.URLParam(r, "token_id"), act.user.ID)
	writeResult(w, map[string]any{"bot_token": token}, err)
}

func (s *Server) publishBotDeleted(deleted store.DeletedBot) {
	for _, workspaceID := range deleted.WorkspaceIDs {
		s.hub.Publish(store.Event{
			Type:        "bot.deleted",
			WorkspaceID: workspaceID,
			CreatedAt:   deleted.DeletedAt,
			Payload: map[string]string{
				"bot_user_id":   deleted.ID,
				"display_name":  deleted.DisplayName,
				"former_handle": deleted.FormerHandle,
				"deleted_at":    deleted.DeletedAt,
			},
		})
	}
}

func (s *Server) publishBotMembershipRemoved(workspaceID, botUserID string) {
	s.hub.Publish(store.Event{
		Type:        "bot.membership_removed",
		WorkspaceID: workspaceID,
		CreatedAt:   time.Now().UTC().Format(time.RFC3339Nano),
		Payload: map[string]string{
			"bot_user_id": botUserID,
		},
	})
}
